api/controllers/OpenEAppController.js

```diff
@@ -150,6 +151,42 @@ const OpenEAppController = {
 
         return expired;
     },
+
+    async downloadReceipt(req, res) {
+        try {
+            await OpenEAppController._streamReceiptToClient(req, res);
+        } catch (err) {
+            sails.log.error(err);
+            return res.serverError();
+        }
+    },
+
+    async _streamReceiptToClient(req, res) {
+        try {
+            OpenEAppController._errorChecks(req, res);
+            sails.log.info('Downloading receipt from Casebook');
+
+            const streamFinished = util.promisify(stream.finished);
+            const response = await CasebookService.getApplicationReceipt(
+                req.params.applicationRef
+            );
+            response.data.pipe(res);
+
+            return streamFinished(res);
+        } catch (err) {
+            throw new Error(`downloadReceipt Error: ${err}`);
+        }
+    },
+
+    _errorChecks(req, res) {
+        const userData = HelperService.getUserData(req, res);
+        if (!userData.loggedIn) {
+            throw new Error('User is not logged in');
+        }
+        if (req.params.applicationRef === 'undefined') {
```

Comment:
is applicationRef === 'undefined'? or typeof applicationRef === 'undefined'?

Reply:
This is down to the wierness of ejs. SO this is part of the code that triggers the route:
```ejs
<a href="/download-receipt/<%= applicationId %>"
```
If `applicationId` is unefined it will return it in this url string so you get the following:
<img width="497" alt="Screenshot 2021-11-25 at 18 12 37" src="https://user-images.githubusercontent.com/1377253/143486916-26a4b98f-595f-48a5-a3a1-0f870e19f517.png">

And so it ends up as a string in the js code:
<img width="598" alt="Screenshot 2021-11-25 at 18 12 48" src="https://user-images.githubusercontent.com/1377253/143486943-8dd0e447-6710-47b6-82f1-3289d6cef739.png">

Reply:
ah I see, what's happening is something like this `download-receipt/${ref}` - ref is undefined, but if you wrap it in a string, it will render `download-receipt/undefined`. 

a couple things.. 

1. what are the cases that will cause applicationId to be undefined
  - also, would it be worth adding an if statement so people aren't greeted with an error page? (only render when applicationId is defined) 
2. I'm assuming this is all auth'd? what is the risk that I navigate to `/download-receipt/random-ref` and download someone else's receipt? is there any risk in that? (am I downloading someone else's data?)